From 36cdc03058b58f77760db82da0d66a0830aeeaa0 Mon Sep 17 00:00:00 2001
From: Timo Rothenpieler <timo@rothenpieler.org>
Date: Sat, 18 May 2024 23:20:46 +0200
Subject: [PATCH 06/21] UPSTREAM: avformat/flvdec: add support for reading
 multi track audio

(cherry picked from commit 25faaa311a74efdfdc4fed56996d7338ed807488)
Signed-off-by: Kexy Biscuit <kexybiscuit@aosc.io>
---
 libavformat/flvdec.c | 21 ++++++++++++++++++---
 1 file changed, 18 insertions(+), 3 deletions(-)

diff --git a/libavformat/flvdec.c b/libavformat/flvdec.c
index f93d7fa75f06..27e243cf0276 100644
--- a/libavformat/flvdec.c
+++ b/libavformat/flvdec.c
@@ -1337,12 +1337,26 @@ retry:
             pkt_type = flags & ~FLV_AUDIO_CODECID_MASK;
 
             if (pkt_type == AudioPacketTypeMultitrack) {
-                av_log(s, AV_LOG_ERROR, "Multitrack audio is unsupported!\n");
-                return AVERROR_PATCHWELCOME;
+                uint8_t types = avio_r8(s->pb);
+                int multitrack_type = types >> 4;
+                pkt_type = types & 0xF;
+
+                if (multitrack_type != MultitrackTypeOneTrack) {
+                    av_log(s, AV_LOG_ERROR, "Audio multitrack types other than MultitrackTypeOneTrack are unsupported!\n");
+                    return AVERROR_PATCHWELCOME;
+                }
+
+                multitrack = 1;
+                size--;
             }
 
             codec_id = avio_rb32(s->pb);
             size -= 4;
+
+            if (multitrack) {
+                track_idx = avio_r8(s->pb);
+                size--;
+            }
         }
     } else if (type == FLV_TAG_TYPE_VIDEO) {
         stream_type = FLV_STREAM_TYPE_VIDEO;
@@ -1439,7 +1453,8 @@ skip:
         st = s->streams[i];
         if (stream_type == FLV_STREAM_TYPE_AUDIO) {
             if (st->codecpar->codec_type == AVMEDIA_TYPE_AUDIO &&
-                (s->audio_codec_id || flv_same_audio_codec(st->codecpar, flags, codec_id)))
+                (s->audio_codec_id || flv_same_audio_codec(st->codecpar, flags, codec_id)) &&
+                st->id == track_idx)
                 break;
         } else if (stream_type == FLV_STREAM_TYPE_VIDEO) {
             if (st->codecpar->codec_type == AVMEDIA_TYPE_VIDEO &&
-- 
2.48.1.windows.1

